File: zo/zo.java
import java.io.*;

public class zo
{
    private static class zavod_t
    {
        int idx; // offset 0x0
        byte[] name_boxed; // offset 0x8
        long name_len; // offset 0x10
        int num_workers; // offset 0x18
        int wtf; // offset 0x1c
        int build_year; // offset 0x20
        zavod_t next_boxed; // offset 0x28
    }
    private static int nz = 0;
    private static zavod_t z_head = null;
    private static boolean _in_have_buf = false;
    private static int _in_buf;
    private static boolean _success;
    private static int getchar() throws Exception
    {
        if(_in_have_buf)
        {
            _in_have_buf = false;
            return _in_buf;
        }
        return System.in.read();
    }
    private static long _read_num() throws Exception
    {
        long ans = 0;
        int c = getchar();
        _success = false;
        while(c == 32 || c == 9 || c == 10)
            c = getchar();
        while(c >= 48 && c < 58)
        {
            _success = true;
            ans = 10 * ans + c - 48;
            c = getchar();
        }
        _in_have_buf = true;
        _in_buf = c;
        return ans;
    }
    private static long read(InputStream fd, byte[] buf, long count) throws Exception
    {
        for(long i = 0; i < count; i++)
        {
            int c = getchar();
            if(c < 0)
                return i;
            buf[(int)i] = (byte)c;
        }
        return count;
    }
    private static void create_z() throws Exception
    {
        if(nz > 10)
            System.exit(1);
        System.out.print("name size: ");
        System.out.flush();
        long name_len = _read_num(); // -0x28(%rbp), 8 bytes
        getchar();
        if(name_len == 0 || name_len >= 0x7ff)
            System.exit(1);
        byte[] name_s = new byte[(int)(name_len + 1)]; // -0x20(%rbp)
        if(name_s == null)
            System.exit(1);
        System.out.print("enter name: ");
        System.out.flush();
        if(read(System.in, name_s, name_len) <= 0)
            System.exit(1);
        name_s[(int)name_len] = 0;
        zavod_t the_zavod = new zavod_t(); // -0x18(%rbp)
        if(the_zavod == null)
            System.exit(1);
        System.out.print("enter number of workers: ");
        System.out.flush();
        the_zavod.num_workers = (int)_read_num();
        getchar();
        System.out.print("is Zavod working (y/N): ");
        System.out.flush();
        int wtf = 0; // -0x10(%rbp), 4 bytes
        byte[] wtf_s = new byte[4];
        read(System.in, wtf_s, 4);
        for(int i = 3; i >= 0; i--)
            wtf = 256 * wtf + (((int)wtf_s[i])&255);
        System.out.print("enter year Zavod was built: ");
        System.out.flush();
        the_zavod.build_year = (int)_read_num();
        if(!_success)
            System.exit(1);
        getchar();
        the_zavod.wtf = 0;
        the_zavod.name_boxed = name_s;
        the_zavod.name_len = name_len;
        the_zavod.idx = nz++;
        System.out.print("Zavod number ");
        System.out.print(the_zavod.idx);
        System.out.println(" has been created\n");
        the_zavod.next_boxed = z_head;
        z_head = the_zavod;
    } 
    private static void edit_z() throws Exception
    {
        System.out.print("enter idx: ");
        System.out.flush();
        int idx = (int)_read_num(); // -0x24(%rbp)
        getchar();
        zavod_t the_zavod = z_head; // -0x20(%rbp)
        while(the_zavod != null && the_zavod.idx != idx)
            the_zavod = the_zavod.next_boxed;
        if(the_zavod == null || the_zavod.name_len == 0)
        {
            System.out.println("no such idx");
            return;
        }
        System.out.print("enter name: ");
        System.out.flush();
        byte[] name_s = the_zavod.name_boxed; // -0x18(%rbp)
        if(read(System.in, name_s, the_zavod.name_len) <= 0)
            System.exit(1);
        name_s[(int)the_zavod.name_len] = 0;
        System.out.print("enter number of workers: ");
        System.out.flush();
        the_zavod.num_workers = (int)_read_num();
        getchar();
        System.out.print("is z working (y/N): ");
        System.out.flush();
        int wtf = 0; // -0x10(%rbp)
        byte[] wtf_s = new byte[4];
        read(System.in, wtf_s, 4);
        for(int i = 3; i >= 0; i--)
            wtf = 256 * wtf + (((int)wtf_s[i])&255);
        the_zavod.wtf = 0;
        System.out.print("enter year Zavod was built: ");
        System.out.flush();
        the_zavod.build_year = (int)_read_num();
        if(!_success)
            System.exit(1);
        getchar();
    }
    private static void print_z() throws Exception
    {
        System.out.print("enter idx: ");
        System.out.flush();
        int idx = (int)_read_num(); // -0x14(%rbp)
        getchar();
        zavod_t the_zavod = z_head; // -0x10(%rbp)
        while(the_zavod != null && the_zavod.idx != idx)
            the_zavod = the_zavod.next_boxed;
        if(the_zavod == null || the_zavod.name_len == 0)
        {
            System.out.println("no such idx");
            return;
        }
        System.out.print("idx: ");
        System.out.println(the_zavod.idx);
        System.out.print("name: ");
        System.out.write(the_zavod.name_boxed, 0, (int)the_zavod.name_len);
        System.out.print("year built: ");
        System.out.println(the_zavod.build_year);
        System.out.print("is working: ");
        System.out.println((the_zavod.wtf!=0)?"Y":"N");
        System.out.print("number of workers: ");
        System.out.println(the_zavod.num_workers);
        System.out.println("=====================================================");
    }
    private static void delete_z() throws Exception
    {
        System.out.print("enter idx: ");
        System.out.flush();
        int idx = (int)_read_num(); // -0x14(%rbp)
        getchar();
        zavod_t the_zavod = z_head; // -0x10(%rbp)
        while(the_zavod != null && the_zavod.idx != idx)
            the_zavod = the_zavod.next_boxed;
        if(the_zavod == null)
            return;
        the_zavod.name_len = 0;
    }
    private static void print_welcome() throws Exception
    {
        System.out.println("+======================================+\n|Welcome to the ZAVOD storing platform!|\n| Here you can manage Zavods you have! |\n+======================================+");
    }
    private static void print_menu() throws Exception
    {
        System.out.println("1. Create Zavod");
        System.out.println("2. Edit Zavod");
        System.out.println("3. Delete Zavod");
        System.out.println("4. Print Zavod");
        System.out.println("5. Exit");
        System.out.print("> ");
        System.out.flush();
    }
    private static void init_stuff() throws Exception
    {
        FileInputStream fd = new FileInputStream("/dev/urandom"); // -0x4(%rbp)
        byte[] cookie = new byte[8];
        if(read(fd, cookie, 8) != 8)
            System.exit(1);
    }
    private static int main() throws Exception
    {
        print_welcome();
        for(;;)
        {
            print_menu();
            int cmd = (int)_read_num(); // -0xc(%rbp)
            if(!_success)
                return 0;
            getchar();
            switch(cmd)
            {
            case 1: create_z(); break;
            case 2: edit_z(); break;
            case 3: delete_z(); break;
            case 4: print_z(); break;
            default: System.exit(0);
            }
        }
    }
    public static void main(String[] argv)
    {
        try
        {
            System.exit(main());
        }
        catch(Exception e)
        {
            e.printStackTrace();
            System.exit(1);
        } 
    }
}

File: zo/zo.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>

typedef struct
{
    int idx; // offset 0x0
    char padding[0x4];
    intptr_t name_boxed; // offset 0x8
    size_t name_len; // offset 0x10
    int num_workers; // offset 0x18
    int wtf; // offset 0x1c
    int build_year; // offset 0x20
    char padding2[0x4];
    intptr_t next_boxed; // offset 0x28
} __attribute__((packed)) zavod_t;


intptr_t cookie;
int nz;
zavod_t* z_head;

void create_z()
{
    if(nz > 10)
        exit(1);
    printf("name size: ");
    size_t name_len; // -0x28(%rbp), 8 bytes
    if(scanf("%zu", &name_len) != 1)
        exit(1);
    getchar();
    if(name_len == 0 || name_len >= 0x7ff)
        exit(1);
    char* name_s = malloc(name_len + 1); // -0x20(%rbp)
    if(name_s == 0)
        exit(1);
    printf("enter name: ");
    if(read(0, name_s, name_len) <= 0)
        exit(1);
    name_s[name_len] = 0;
    zavod_t* the_zavod = malloc(0x30); // -0x18(%rbp)
    if(the_zavod == 0)
        exit(1);
    printf("enter number of workers: ");
    scanf("%d", &the_zavod->num_workers);
    getchar();
    printf("is Zavod working (y/N): ");
    int wtf; // -0x10(%rbp), 4 bytes
    read(0, &wtf, 4);
    printf("enter year Zavod was built: ");
    if(scanf("%d", &the_zavod->build_year) != 1)
        exit(1);
    getchar();
    the_zavod->wtf = 0;
    the_zavod->name_boxed = ((intptr_t)name_s) ^ cookie;
    the_zavod->name_len = name_len;
    the_zavod->idx = nz++;
    printf("Zavod number %d has been created\n", the_zavod->idx);
    the_zavod->next_boxed = ((intptr_t)the_zavod)^((intptr_t)z_head)^cookie;
    z_head = the_zavod;
}

void edit_z()
{
    printf("enter idx: ");
    int idx; // -0x24(%rbp)
    scanf("%d", &idx);
    getchar();
    zavod_t* the_zavod = z_head; // -0x20(%rbp)
    while(the_zavod && the_zavod->idx != idx)
        the_zavod = (zavod_t*)(the_zavod->next_boxed^((intptr_t)the_zavod)^cookie);
    if(!the_zavod || the_zavod->name_len == 0)
    {
        puts("no such idx");
        return;
    }
    printf("enter name: ");
    char* name_s = (char*)(the_zavod->name_boxed^cookie); // -0x18(%rbp)
    if(read(0, name_s, the_zavod->name_len) <= 0)
        exit(1);
    name_s[the_zavod->name_len] = 0;
    printf("enter number of workers: ");
    scanf("%d", &the_zavod->num_workers);
    getchar();
    printf("is z working (y/N): ");
    int wtf; // -0x10(%rbp)
    read(0, &wtf, 4);
    the_zavod->wtf = 0;
    printf("enter year Zavod was built: ");
    if(scanf("%d", &the_zavod->build_year) != 1)
        exit(1);
    getchar();
}

void print_z()
{
    printf("enter idx: ");
    int idx; // -0x14(%rbp)
    scanf("%d", &idx);
    getchar();
    zavod_t* the_zavod = z_head; // -0x10(%rbp)
    while(the_zavod && the_zavod->idx != idx)
        the_zavod = (zavod_t*)(the_zavod->next_boxed^((intptr_t)the_zavod)^cookie);
    if(!the_zavod || the_zavod->name_len == 0)
    {
        puts("no such idx");
        return;
    }
    printf("idx: %d\n", the_zavod->idx);
    printf("name: ");
    write(1, (char*)(the_zavod->name_boxed^cookie), the_zavod->name_len);
    printf("year built: %d\n", the_zavod->build_year);
    printf("is working: %s\n", (the_zavod->wtf)?"Y":"N");
    printf("number of workers: %d\n", the_zavod->num_workers);
    puts("=====================================================");
}

void delete_z()
{
    printf("enter idx: ");
    int idx; // -0x14(%rbp)
    scanf("%d", &idx);
    getchar();
    zavod_t* the_zavod = z_head; // -0x10(%rbp)
    while(the_zavod && the_zavod->idx != idx)
        the_zavod = (zavod_t*)(the_zavod->next_boxed^((intptr_t)the_zavod)^cookie);
    if(the_zavod == 0)
        return;
    free((void*)(the_zavod->name_boxed^cookie));
    the_zavod->name_len = 0;
}

void print_welcome()
{
    puts("+======================================+\n|Welcome to the ZAVOD storing platform!|\n| Here you can manage Zavods you have! |\n+======================================+");
}

void print_menu()
{
    puts("1. Create Zavod");
    puts("2. Edit Zavod");
    puts("3. Delete Zavod");
    puts("4. Print Zavod");
    puts("5. Exit");
    printf("> ");
}

void init_stuff()
{
    int fd = open("/dev/urandom", 0); // -0x4(%rbp)
    if(fd < 0)
        exit(1);
    if(read(fd, &cookie, 8) != 8)
        exit(1);
    close(fd);
}

int main()
{
    setvbuf(stdin, 0, 2, 0);
    setvbuf(stdout, 0, 2, 0);
    alarm(0x1e);
    print_welcome();
    for(;;)
    {
        print_menu();
        int cmd; // -0xc(%rbp)
        if(scanf("%d", &cmd) != 1)
            exit(0);
        getchar();
        switch(cmd)
        {
        case 1: create_z(); break;
        case 2: edit_z(); break;
        case 3: delete_z(); break;
        case 4: print_z(); break;
        default: exit(0);
        }
    }
}
